File: src/1_100/0092_Reverse_Linked_List_II/Reverse_Linked_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
    	if(head==null || m<=0 || n<=0 || m>=n) {
    		return head;
    	}
    	ListNode dummy = new ListNode(-1);
    	ListNode p = dummy;
    	p.next = head;
    	int i = 0;
    	while(p!=null && i<m-1) {
    		p = p.next;
    		i++;
    	}
    	if(p == null) {
    		return head;
    	}
    	LinkedList<ListNode> stack = new LinkedList<ListNode>();
    	ListNode cur = p.next;
    	ListNode next = null;
    	while(cur!=null && i<n) {
    		stack.push(cur);
    		next = cur.next;
    		cur = cur.next;
    		i++;
    	}
    	p.next = null;
    	while(!stack.isEmpty()) {
    		p.next = stack.pop();
    		p = p.next;
    		p.next = null;
    	}
    	p.next = null;
    	p.next = next;
    	return dummy.next;
    }
}

File: src/1_100/0092_Reverse_Linked_List_II/Reverse_Linked_List_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
	# 先找到反转字符串的前一个str_pre，然后根据 k<n，不断反转链表
	# 最后将str_pre->next = pre
	# tmp是反转链表的后面一个节点，str_tail是反转链表后的尾巴
	# 将 str_tail->next = tmp 整个链表就 串联起来了
    def reverseBetween(self, head, m, n):
        p = ListNode(-1)
        cur,k,p.next = p,0,head
        while cur and k<m-1:
            cur,k = cur.next,k+1
        str_pre,str_tail = cur,cur.next
        str_pre.next,pre,tmp,cur = None,None,None,str_tail
        while cur and k<n:
            tmp,cur.next,pre,cur,k = cur.next,pre,cur,cur.next,k+1
        str_pre.next,str_tail.next = pre,tmp
        return p.next